feat: add table_hive_ha materialization

nicor88 opened the pull request:
### Description

Add a new materialisation called `table_hive_ha` that extend the create_table_as, but with a not destructive behavior.

If the table is created for the first time it works like a normal table materialization.
If the table already exist, it create a tmp table (with the same attribute as the target relation), and then it uses a glue table swap that just point the target table to the new tmp table location.
If everything is successful the old table location is pruned, and the tmp table is dropped (note only catalog/glue definition).

To test that the materialization was working as expected I used this piece of code:
```
from pyathena import connect
import boto3
import os

os.environ['AWS_PROFILE'] = 'bi-staging'

session = boto3.session.Session(region_name='eu-central-1')

cursor = connect(
    work_group='data-engineering',
    session=session
).cursor()

while True:
    try:
        cursor.execute("select count(*) from silver_shared.ha_1")
        result = cursor.fetchall()
        print(result)
    except Exception as err:
        print(err)
```

## Remarks
*~ I add to refactor some boto3 calls on the way, as we were leaving so many clients open~ Refactor cover in another PR
* I notice a super little downtime when swapping from a table without partitions with a table with partitions, and reverse.

# TODO
- [x] Handle partitions and change of table with partitions
- [x] Add unit testing for swap_table
- [ ] Add unit testing for expire_glue_table_versions

## Models used to test - Optional
### With partitions
```
{{ config(
    materialized='table_hive_ha',
    format='parquet',
    s3_data_naming='table',
    partitioned_by=['status']
) }}

select
    'a' as user_id,
    'pi' as user_name,
    'active' as status
union all
select
    'b' as user_id,
    'sh' as user_name,
    'active' as status
union all
select
    'c' as user_id,
    'sh' as user_name,
    'disabled' as status
```

### Without partitions
```
{{ config(
    materialized='table_hive_ha',
    format='parquet',
    s3_data_naming='table'
) }}

select
    'a' as user_id,
    'pi' as user_name,
    'active' as status
union all
select
    'b' as user_id,
    'sh' as user_name,
    'active' as status
union all
select
    'c' as user_id,
    'sh' as user_name,
    'disabled' as status

```
## Checklist
- [ ] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [ ] You added unit testing when necessary
- [ ] You added functional testing when necessary

juliansteger-sc commented:
This is great work, so first thanks @nicor88 for contributing so much to this adapter!

We're working on a similar feature, and I just want to point out some differences / issues:
- I'm not sure how schema evolution is handled in your solution. In our case, we compare schema of the tmp table and the original one, if it changes, we still drop and recreate the table (a more advanced solution could add / remove columns as needed, as described [here](https://docs.aws.amazon.com/athena/latest/ug/handling-schema-updates-chapter.html))
- We just extended the `schema_table_unique` and `table_unique` options to have the zero-downtime option as default, and the "old behavior" can be triggered with `--full-refresh` option.
- When pruning the old s3 location, it could be there are still user queries running, that need the data in the old location I think. We solved this by having a configurable timeout, before location is pruned (a more advanced solution would be to do a nightly cleanup job for stale data)
- you have some logging in your `swap_table` that doesn't seem to work with `--log-format json` (json logging)

I could open a PR as soon we have finished the implementation, not sure when you plan to merge this.

nicor88 commented:
> I'm not sure how schema evolution is handled in your solution. In our case, we compare schema of the tmp table and the original one, if it changes, we still drop and recreate the table (a more advanced solution could add / remove columns as needed, as described [here](https://docs.aws.amazon.com/athena/latest/ug/handling-schema-updates-chapter.html))
the schema evolution is totally covered, as we leverage table updates in glue, that support totally different schemas.

It's possible to also switch from a table with partitions to a table without.

nicor88 commented:
@juliansteger-sc @hiltercoty  have a look again.

Instead of immediately deleting the old path, I added a function called: **expire_glue_table_versions**  that keep the last x versions of a table.
Note that when calling `update_table` glue add 2 versions, so for now I kept the last 4 versions of a table by default.

To test his feature I had an always running query in a while True loop (see PR description), and I didn't see any failures when switching from table to table.

When switching from table with partitions to a table without I saw 1 query failing, as the partitions need to be reset, and the table is inconsistent from a fraction of time (<1sec).

>you have some logging in your swap_table that doesn't seem to work with --log-format json (json logging)
@juliansteger-sc I added messages only in debug mode, and I think in the overall adapter we don't use JSON formatting. But happy to have exactly what is breaking to fix it.

> I could open a PR as soon we have finished the implementation, not sure when you plan to merge this.
I'm happy to take this in private or here, and agree on what are the most important features to allign with, to have a general community approach and avoid many version of the same solution around.
In case feel fee to ping me in dbt Slack :)

Also, this PR is experimental, me and @Jrmyy agreed to have the less breaking changes in the create_table_as when introducing this feature.

juliansteger-sc commented:
makes all sense, and nice idea with the `expire_glue_table_versions`.
Regarding logging, I think json logging is a dbt feature: https://docs.getdbt.com/reference/global-configs#log-formatting
We use it in production because makes log querying easier.

Maybe you can try if it works without exception on your end. (`dbt --log-format json run`, or even with `--debug`)

nicor88 commented:
@juliansteger-sc thanks for the hint, I was not using the json logging, that could be indeed super powerfull (based on the logging system in use).
The logging issue should be fixed now, it was due to boto3 return datetime objects, as json.dumps, used by the logging don't know how to treat datetimes (generally it's possible to pass a custom serializer to json.dumps) I decided to remove the full boto3 responses and add a more `custom made` logging of the steps that are executed under the hood.

nicor88 commented:
@juliansteger-sc did you manage to test this feature?

nicor88 commented:
Some refactors are implemented in this PR: https://github.com/dbt-athena/dbt-athena/pull/141

Jrmyy commented on tests/unit/test_adapter.py:
```diff
@@ -546,6 +546,109 @@ def test_list_relations_without_caching_with_non_glue_data_catalog(self, parent_
+    @mock_athena
+    @mock_glue
+    @mock_s3
+    def test_swap_table_with_no_partitions_to_one_without(self, aws_credentials):
```
Is it `test_swap_table_with_no_partitions_to_one_with` ?

nicor88 replied:
yes it is, good catch!

nicor88 commented on tests/unit/test_adapter.py:
```diff
@@ -546,6 +546,109 @@ def test_list_relations_without_caching_with_non_glue_data_catalog(self, parent_
+        table_versions = glue.get_table_versions(DatabaseName=DATABASE_NAME, TableName=table_name).get("TableVersions")
+        print(table_versions)
+        assert len(table_versions) == 3
+        # TODO due to a limitation with moto this is not fully testable
+        # self.adapter.expire_glue_table_versions(DATABASE_NAME, table_name, 1, False)
+        # table_versions_after_expiration = glue.get_table_versions(
+        #     DatabaseName=DATABASE_NAME,
+        #     TableName=table_name).get("TableVersions")
+        # assert len(table_versions_after_expiration) == 1
```
@Jrmyy this is the big blocker so far.
I'm missing the VersionId and the UpdatedTime in the returned object from moto....

nicor88 commented:
CI will be fixed when this is merged: https://github.com/dbt-athena/dbt-athena/pull/150/files

ignacioreyna commented:
Hello! This is an awesome feature, thanks for developing it! I'm really interested in it so I was just testing it but encountered an issue, though I'm not sure if it's something with my Glue config.

I was trying to create a model `test.sql` within the schema `migration_test` with the following snippet

```
{{ config(
    materialized='table_hive_ha',
    s3_data_naming='schema_table_unique'
) }}

select 
    2 as inte
    , cast('2022-12-12 12:12:12' as timestamp) as ts
    , 'asd' as str
    , 2.2 as double
    , 1=1 as bool

```

But that would throw me an error which seemed weird.

`botocore.errorfactory.InvalidRequestException: An error occurred (InvalidRequestException) when calling the StartQueryExecution operation: line 1:22: mismatched input '"migration_test"' expecting `

So when I tried to find out why this could be I remembered seeing this error whenever I tried to execute a _Hive_ query in Athena using double quotes instead of backticks.

I was able to fix it by adding a `.render_hive()` to the `tmp_relation`s in the DDL queries within the new `table_hive_ha.sql` (I just copied that from the `relation.sql` in the `adapters` dir!). Will create a new PR now!

nicor88 commented:
@ignacioreyna thanks for spotting that, it's indeed related to the new hive quoting support added recently.
